File: 08-安全/01-密码学/对称加密/AES-GCM/go/main.go
// AES-128-GCM demo — 使用 Go 标准库 crypto/aes + crypto/cipher GCM
//
// 参考:
//   NIST SP 800-38D, "Recommendation for Block Cipher Modes of Operation:
//   Galois/Counter Mode (GCM) and GMAC", November 2007
//   https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38d.pdf
//   Go 标准库 crypto/cipher.NewGCM 实现即遵循 NIST 800-38D §7
//
// 运行: go run main.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/hex"
	"fmt"
)

// hex2b: 把 hex 字符串解码成字节切片,出错直接 panic(测试向量固定,不应出错)。
func hex2b(s string) []byte {
	b, err := hex.DecodeString(s)
	if err != nil {
		panic(err)
	}
	return b
}

// eq: 比较两字节切片是否相等,不等则格式化报错。
func eq(name string, got, want []byte) {
	if hex.EncodeToString(got) != hex.EncodeToString(want) {
		panic(fmt.Sprintf("%s 失配:\n  got  = %x\n  want = %x", name, got, want))
	}
	fmt.Printf("  ✓ %s\n", name)
}

// runAEAD: 调用 stdlib GCM 完成 encrypt/decrypt,并验证 NIST 测试向量。
//
// AES-GCM 的内部流程(NIST SP 800-38D §7):
//   1. J0 = IV || 0x00000001(IV=12B 时 deterministic construction)
//   2. H = E_K(0^128),作为 GHASH 子密钥
//   3. C = GCTR_K(J0+1, P)         — 计数模式加密
//   4. S = GHASH_H(A || C || lenA*8 || lenC*8)
//   5. T = GCTR_K(J0, S)[:t]        — 认证 tag
//
// crypto/cipher.NewGCM 内部即按此实现;这里只演示其外部 API 行为。
func runAEAD(block cipher.Block, iv, pt, aad, wantCT, wantTag []byte, label string) {
	aead, err := cipher.NewGCM(block)
	if err != nil {
		panic(err)
	}
	// Seal 把 AAD、nonce、plaintext 一起喂给 GCM,返回 ciphertext || tag
	ctWithTag := aead.Seal(nil, iv, pt, aad)
	ct := ctWithTag[:len(pt)]
	tag := ctWithTag[len(pt):]
	eq(label+" ct", ct, wantCT)
	eq(label+" tag", tag, wantTag)

	// Open:验证 tag + 解密回 plaintext(常量时间比较在 stdlib 内部完成)
	ptBack, err := aead.Open(nil, iv, ctWithTag, aad)
	if err != nil {
		panic(fmt.Sprintf("%s Open 失败: %v", label, err))
	}
	eq(label+" decrypt round-trip", ptBack, pt)
}

func main() {
	fmt.Println("[1] AES-128-GCM KAT (NIST SP 800-38D Appendix B)")

	// Test Case 1:全零 key/iv,空 pt + 空 aad
	{
		block, _ := aes.NewCipher(make([]byte, 16))
		aead, _ := cipher.NewGCM(block)
		ctWithTag := aead.Seal(nil, make([]byte, 12), []byte(""), []byte(""))
		eq("TC1 ct (空)", ctWithTag[:0], []byte{})
		eq("TC1 tag", ctWithTag, hex2b("58e2fccefa7e3061367f1d57a4e7455a"))
	}

	// Test Case 2:全零 key/iv,16B 零 pt,空 aad
	{
		block, _ := aes.NewCipher(make([]byte, 16))
		runAEAD(block,
			make([]byte, 12), make([]byte, 16), nil,
			hex2b("0388dace60b6a392f328c2b971b2fe78"),
			hex2b("ab6e47d42cec13bdf53a67b21257bddf"),
			"TC2")
	}

	// Test Case 3:64B pt,空 aad
	{
		block, _ := aes.NewCipher(hex2b("feffe9928665731c6d6a8f9467308308"))
		runAEAD(block,
			hex2b("cafebabefacedbaddecaf888"),
			hex2b("d9313225f88406e5a55909c5aff5269a86a7a9531534f7da2e4c303d8a318a72"+
				"1c3c0c95956809532fcf0e2449a6b525b16aedf5aa0de657ba637b391aafd255"),
			nil,
			hex2b("42831ec2217774244b7221b784d0d49ce3aa212f2c02a4e035c17e2329aca12e"+
				"21d514b25466931c7d8f6a5aac84aa051ba30b396a0aac973d58e091473f5985"),
			hex2b("4d5c2af327cd64a62cf35abd2ba6fab4"),
			"TC3")
	}

	// Test Case 4:60B pt,20B AAD
	{
		block, _ := aes.NewCipher(hex2b("feffe9928665731c6d6a8f9467308308"))
		runAEAD(block,
			hex2b("cafebabefacedbaddecaf888"),
			hex2b("d9313225f88406e5a55909c5aff5269a86a7a9531534f7da2e4c303d8a318a72"+
				"1c3c0c95956809532fcf0e2449a6b525b16aedf5aa0de657ba637b39"),
			hex2b("feedfacedeadbeeffeedfacedeadbeefabaddad2"),
			hex2b("42831ec2217774244b7221b784d0d49ce3aa212f2c02a4e035c17e2329aca12e"+
				"21d514b25466931c7d8f6a5aac84aa051ba30b396a0aac973d58e091"),
			hex2b("5bc94fbc3221a5db94fae95ae7121a47"),
			"TC4")
	}

	// [2] 篡改检测:Open 在 tag/AAD/CT 任一被改时返回 error
	fmt.Println("\n[2] AES-GCM 篡改检测")
	block, _ := aes.NewCipher(hex2b("feffe9928665731c6d6a8f9467308308"))
	aead, _ := cipher.NewGCM(block)
	iv := hex2b("cafebabefacedbaddecaf888")
	pt := []byte("hello world")
	aad := []byte("meta")
	ctWithTag := aead.Seal(nil, iv, pt, aad)
	bad := append([]byte(nil), ctWithTag...)
	bad[0] ^= 0x01 // 翻一个密文位
	if _, err := aead.Open(nil, iv, bad, aad); err == nil {
		panic("篡改密文未被检测")
	} else {
		fmt.Println("  ✓ 密文位翻转被检测")
	}
	// 改 AAD
	if _, err := aead.Open(nil, iv, ctWithTag, []byte("metA")); err == nil {
		panic("篡改 AAD 未被检测")
	} else {
		fmt.Println("  ✓ AAD 改动被检测")
	}
	// 改 tag 末字节
	badTag := append([]byte(nil), ctWithTag...)
	badTag[len(badTag)-1] ^= 0x01
	if _, err := aead.Open(nil, iv, badTag, aad); err == nil {
		panic("篡改 tag 未被检测")
	} else {
		fmt.Println("  ✓ tag 改动被检测")
	}

	fmt.Println("\n全部 AES-GCM 测试通过 ✓")
}

File: 08-安全/01-密码学/对称加密/AES-GCM/python/main.py
"""AES-GCM demo 入口。

按 NIST SP 800-38D Appendix B / 测试用例 1-4 验证自实现与官方测试向量一致。

运行:
    python3 main.py
"""

import sys
from pathlib import Path

# 允许 `python main.py` 直接运行(把当前目录加到 sys.path)
sys.path.insert(0, str(Path(__file__).parent))

from aes128 import Aes128  # noqa: E402
from gcm import AesGcm  # noqa: E402


def hex2b(h: str) -> bytes:
    return bytes.fromhex(h)


def assert_eq(name: str, got: bytes, want: bytes) -> None:
    if got != want:
        raise AssertionError(f"{name} 失配:\n  got  = {got.hex()}\n  want = {want.hex()}")
    print(f"  ✓ {name}")


def test_aes128_kat() -> None:
    """FIPS 197 Appendix B 的 AES-128 已知答案测试。"""
    print("[1] AES-128 KAT (FIPS 197 Appendix B)")
    # 全部为零的 key + pt,期望 ct = 66 e9 4b d4 ef 8a 2c 3b 88 4c fa 59 ca 34 2b 2e
    cipher = Aes128(b"\x00" * 16)
    pt = b"\x00" * 16
    ct = cipher.encrypt_block(pt)
    assert_eq("FIPS197 B.1 加密",
              ct, hex2b("66e94bd4ef8a2c3b884cfa59ca342b2e"))
    # 解密还原
    pt2 = cipher.decrypt_block(ct)
    assert_eq("FIPS197 B.1 解密", pt2, pt)


def test_aes_gcm_kat() -> None:
    """NIST SP 800-38D 推荐的 AES-GCM 测试用例(Appendix B Test Case 1-4)。"""
    print("\n[2] AES-128-GCM KAT (NIST SP 800-38D Appendix B)")

    # Test Case 1: 全零 key/iv,空 pt + 空 aad
    gcm = AesGcm(b"\x00" * 16)
    ct, tag = gcm.encrypt(b"\x00" * 12, b"", b"")
    assert_eq("TC1 ct (空)", ct, b"")
    assert_eq("TC1 tag", tag, hex2b("58e2fccefa7e3061367f1d57a4e7455a"))

    # Test Case 2: 全零 key/iv,16 字节零 pt,空 aad
    gcm = AesGcm(b"\x00" * 16)
    ct, tag = gcm.encrypt(b"\x00" * 12, b"\x00" * 16, b"")
    assert_eq("TC2 ct", ct, hex2b("0388dace60b6a392f328c2b971b2fe78"))
    assert_eq("TC2 tag", tag, hex2b("ab6e47d42cec13bdf53a67b21257bddf"))

    # Test Case 3: 非零 key/iv,64 字节 pt,空 aad
    gcm = AesGcm(hex2b("feffe9928665731c6d6a8f9467308308"))
    pt = hex2b("d9313225f88406e5a55909c5aff5269a"
               "86a7a9531534f7da2e4c303d8a318a72"
               "1c3c0c95956809532fcf0e2449a6b525"
               "b16aedf5aa0de657ba637b391aafd255")
    expected_ct = hex2b("42831ec2217774244b7221b784d0d49c"
                        "e3aa212f2c02a4e035c17e2329aca12e"
                        "21d514b25466931c7d8f6a5aac84aa05"
                        "1ba30b396a0aac973d58e091473f5985")
    expected_tag = hex2b("4d5c2af327cd64a62cf35abd2ba6fab4")
    ct, tag = gcm.encrypt(hex2b("cafebabefacedbaddecaf888"), pt, b"")
    assert_eq("TC3 ct", ct, expected_ct)
    assert_eq("TC3 tag", tag, expected_tag)

    # Test Case 4: 同 TC3 但带 20 字节 AAD,pt 砍到 60 字节
    aad = hex2b("feedfacedeadbeeffeedfacedeadbeefabaddad2")
    pt4 = hex2b("d9313225f88406e5a55909c5aff5269a"
                "86a7a9531534f7da2e4c303d8a318a72"
                "1c3c0c95956809532fcf0e2449a6b525"
                "b16aedf5aa0de657ba637b39")
    expected_ct4 = hex2b("42831ec2217774244b7221b784d0d49c"
                         "e3aa212f2c02a4e035c17e2329aca12e"
                         "21d514b25466931c7d8f6a5aac84aa05"
                         "1ba30b396a0aac973d58e091")
    expected_tag4 = hex2b("5bc94fbc3221a5db94fae95ae7121a47")
    ct, tag = gcm.encrypt(hex2b("cafebabefacedbaddecaf888"), pt4, aad)
    assert_eq("TC4 ct", ct, expected_ct4)
    assert_eq("TC4 tag", tag, expected_tag4)

    # 解密回环测试
    pt4_back = gcm.decrypt(hex2b("cafebabefacedbaddecaf888"), ct, aad, tag)
    assert_eq("TC4 decrypt round-trip", pt4_back, pt4)


def test_aes_gcm_tamper() -> None:
    """篡改 ct / aad / tag 任一项,解密应抛 ValueError。"""
    print("\n[3] AES-GCM 篡改检测")
    gcm = AesGcm(hex2b("feffe9928665731c6d6a8f9467308308"))
    iv = hex2b("cafebabefacedbaddecaf888")
    pt = b"hello world" * 4
    aad = b"meta"
    ct, tag = gcm.encrypt(iv, pt, aad)
    # 改一个密文字节
    bad_ct = bytearray(ct); bad_ct[0] ^= 0x01
    try:
        gcm.decrypt(iv, bytes(bad_ct), aad, tag)
    except ValueError:
        print("  ✓ 密文位翻转被检测")
    else:
        raise AssertionError("篡改密文未触发校验")
    # 改 aad
    try:
        gcm.decrypt(iv, ct, b"metA", tag)
    except ValueError:
        print("  ✓ AAD 改动被检测")
    else:
        raise AssertionError("篡改 AAD 未触发校验")
    # 改 tag
    bad_tag = bytearray(tag); bad_tag[-1] ^= 0x01
    try:
        gcm.decrypt(iv, ct, aad, bytes(bad_tag))
    except ValueError:
        print("  ✓ tag 改动被检测")
    else:
        raise AssertionError("篡改 tag 未触发校验")


def main() -> None:
    test_aes128_kat()
    test_aes_gcm_kat()
    test_aes_gcm_tamper()
    print("\n全部 AES-GCM 测试通过 ✓")


if __name__ == "__main__":
    main()